### PR TITLE
Reusable blocks: prevent infinite recursion

### LINKED ISSUE #18704: 500 error and 'Reusable' tab disappears when a reusable block is embedded into itself
**Describe the bug**
When a reusable block has a reusable block embedded (containing the same ID)  it throws a 500 error and the 'Reusable' tab disappears in WP 5.3.

**To reproduce**
1 Go to  wp-admin/edit.php?post_type=wp_block&paged=1

2. Create a new reusable block. For this test, I am calling it 'Reusable Block Test'. You can call it anything you'd like. Publish and save, note the post ID from the URL.

3. Embed a new reusable block into the one you just created. Pick a random block.

p10

4. Switch the view to 'Code Editor' and enter the ID from step #2. Update the post.

p11

5. You should get an error : 'Updating failed. or Updating failed. Error message: The response is not a valid JSON response.  (note that it did update)

6. Go to Posts -> Add New. Open up the "Network" tab in Google Chrome. Clear anything there to get a fresh list. 

p12

7. Add a new block. You should get a 500 error and the "Reusable" tab is missing. 

p15

Request URL: */wp-json/wp/v2/blocks?per_page=100&_locale=user
Request Method: GET
Status Code: 500 
Remote Address: 127.0.0.1:8080
Referrer Policy: strict-origin-when-cross-origin

p13

8. Editing the reusable block from step 2 results in a frontend error: 
"There has been a critical error on your website. Please check your site admin email inbox for instructions"

**Expected behavior**
Adding a block allows you to add previously created Reusable Blocks by using the "Reusable" tab

**Desktop (please complete the following information):**
 - OS: [e.g. iOS] Windows 10
 - Browser [e.g. chrome, safari]: Chrome
 - Version [e.g. 22] 78.0.3904.108 

**Additional context**
WP Version:  5.3

### PR DESCRIPTION
**Pull request by mcsf**
## Description
Alternative to: #27012, #26951
Fixes: #18704

In contrast with the aforementioned pull requests, this builds on the opinion that any block type capable of recursion should be responsible for its own infinite loops. In other words, it isn't up to the block editor framework to detect and prevent block loops.

* **On the front end.** This is the simplest end of the fix. Thanks to PHP's single-pass execution, it is enough to keep a static array inside the `render_block_core_block`. If a given Reusable Block `ref` is already present in the array, prevent further rendering and issue a PHP warning (`E_USER_WARNING`).

* **On the client.** We resort to React context in order to identify any blocks that have already been rendered in the same render tree. The principle here is that each `ReusableBlockEdit` instance _first reads_ and _then expands_ the context:

```js
const previouslyRenderedRefs = useContext( RenderedRefsContext );
const hasAlreadyRendered = previouslyRenderedRefs.includes( ref );
const newRenderedRefs = useMemo( () => [ ...previouslyRenderedRefs, ref ], [
	ref,
	previouslyRenderedRefs,
] );
// later
return (
	<RenderedRefsContext.Provider value={ newRenderedRefs }>
		{ /* contents go here */ }
```

This behaviour is encapsulated in the `useNoRecursiveRenders` hook, for which tests are provided.

## Notes

* `useNoRecursiveRenders` is general enough that other block types. The Template Part block would be able to use it directly.
* For now, everything lives under `block-library/src/block` because my priority is to prevent fatals reported out there in production, but once this PR is merged we can expand the fix to Template Part.
* Internally, `useNoRecursiveRenders`'s `previouslyRenderedRefs` could be a [`Set`](https://wiki.developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) instead of an `Array`.
* We considered WordPress's own Context API instead of React Context, but the Context API only allows a piece of context to be overwritten by a provider's attribute. This doesn't meet our needs, as we need to keep track of a growing set of values, not a singleton.

## How has this been tested?

* Added specific tests for new hook `useNoRecursiveRender`
* Manual testing involves any of:
  * Creating a Reusable Block that contains itself
  * Creating a pair of Reusable Blocks that contain each other
  * Creating a pair of (Reusable Block + Template Part) that contain each other
  * ... and ensuring neither the front end nor the editor crash.

## Screenshots

Front end | Editor
-|-
Captura de ecrã 2021-01-22, às 12 23 49|Captura de ecrã 2021-01-22, às 09 50 14

## Types of changes

Bug fix

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-editor/src/components/warning/style.scss
Comment (mcsf) [inline packages/block-editor/src/components/warning/style.scss:17]: Don't merge before removing this.

**Comment by carlomanf** [comment]
> In contrast with the aforementioned pull requests, this builds on the opinion that any block type capable of recursion should be responsible for its own infinite loops.

Would your solution also handle infinite loops that involve more than one block type? (e.g. a reusable block and a template part that both include each other)

Tagging @bobbingwide

**Comment by ntsekouras** [comment]
> Would your solution also handle infinite loops that involve more than one block type? (e.g. a reusable block and a template part that both include each other)

I just tested this and seems to work good! I tested with many combinations, including:
1. `reusable` blocks (A inserts A)
2. `reusable` blocks  (A contains B], B inserts A)
3. `template part` (TP contains A reusable, A inserts TP)

File: packages/block-library/src/block/index.php
Comment (ntsekouras) [inline packages/block-library/src/block/index.php:23]: We should change this :)

File: packages/block-editor/src/components/warning/style.scss
Comment (ntsekouras) [reply packages/block-editor/src/components/warning/style.scss:17]: Why you needed `!important` here? I tested with TT1-blocks and wasn't affecting anything?

**Comment by bobbingwide** [comment]
> I just tested this and seems to work well! I tested with many combinations, including:

Did you try?

- Template or reusable block being embedded twice in the same content. 

> (More notes on this PR to come…)
> How has this been tested?

I'd like to see these notes and test results.

File: packages/block-editor/src/components/warning/style.scss
Comment (mcsf) [reply packages/block-editor/src/components/warning/style.scss:17]: By sheer luck, I was using a Group block with a dark background colour, and that is how I noticed that the `.editor-styles-wrapper .has-background ...` rules override Warning's colour:

`color: currentColor` | `color: $gray-900`
-|-
Captura de ecrã 2021-01-22, às 09 50 40|Captura de ecrã 2021-01-22, às 09 50 14

File: packages/block-library/src/block/index.php
Comment (mcsf) [reply packages/block-library/src/block/index.php:23]: I'm glad you bring that up! I'm thinking that, for the front end, we should always show a disruptive warning as we do in the editor. We should probably honour the right constants or settings, e.g. `WP_DEBUG`.

If a site disables warnings, then we should just `return '';`. Otherwise, we can trigger a PHP warning.

Captura de ecrã 2021-01-22, às 10 15 50

My only hesitation is that the PHP warning alone doesn't really help the site admin find the offending block. See the screenshot above. The text that reads _[block rendering halted]_ is returned by `render_block_core_block`.

Comment (mcsf) [reply packages/block-library/src/block/index.php:23]: (edited: _should_ -> _should **not**_ 😅)

Comment (ntsekouras) [reply packages/block-library/src/block/index.php:23]: Your idea seems valid to me and you can just add a better message like this:
```
trigger_error("Reusable block `$reusable_block->post_title`, tries to render itself and can cause infinite loop...", E_USER_WARNING);
```

with some different wording probably :)

Comment (mcsf) [reply packages/block-library/src/block/index.php]: Good idea. Pushed in 9c7d4af1914309aff80bfbd5f2cd6ac5652fa4ab.

**Review by aristath — APPROVED** [review]
This is lovely! Works as advertised and the code is clean and easy to understand.
With the exception of the `!important` in the `.scss` file this one looks good to me :+1:

File: packages/block-library/src/block/index.php
Comment (ntsekouras) [inline packages/block-library/src/block/index.php:28]: Should we also show in in admin? I think the `Warning` is enough there.
Maybe trigger if `! is_admin()`?

File: packages/block-library/src/block/use-no-recursive-renders.js
Comment (gziolo) [inline packages/block-library/src/block/use-no-recursive-renders.js:14]: Not a blocker for this PR. I wanted to share my thought on how to make it general enough to use with other blocks that require the same guard (Post Content or Template Part). It looks like it would be enough to rename `ref` to something more generic like `uniqueId` and use `Set` rather than array of strings for checks to support compound values for more complex blocks.

File: packages/block-editor/src/components/warning/style.scss
Comment (ntsekouras) [reply packages/block-editor/src/components/warning/style.scss:17]: I also added `background` in a group that contains nested reusables and still the color seemed fine 🤔 
I also changed the `color settings` and yet no problem.. What theme are you using?

File: packages/block-library/src/block/use-no-recursive-renders.js
Comment (mcsf) [reply packages/block-library/src/block/use-no-recursive-renders.js:14]: Agreed, thanks for the note!

**Review by gziolo — COMMENTED** [review]
It looks solid, I really like `useNoRecursiveRenders` hook with the accompanying tests at is could be reused with other blocks. Later (after this fix is done), we can consider moving it to the `@wordpress/block-editor` package.

It looks like it only needs some polishing on UI side. Great work @mcsf 👏🏻

File: packages/block-library/src/block/index.php
Comment (mcsf) [reply packages/block-library/src/block/index.php:28]: Good catch! On my end, the warning was hidden by Gutenberg's full screen mode.

Comment (mcsf) [reply packages/block-library/src/block/index.php]: → 90eb5866c4

File: packages/block-editor/src/components/warning/style.scss
Comment (mcsf) [reply packages/block-editor/src/components/warning/style.scss:17]: I'm using _Seedlet (Blocks)_, which hasn't been updated in a while, so that may be the root issue. Thanks for testing with TT1! I'll remove this patch now.

File: packages/block-library/src/block/index.php
Comment (mcsf) [inline packages/block-library/src/block/index.php:39]: Thoughts on leaving this here or returning `''`?

Comment (ntsekouras) [reply packages/block-library/src/block/index.php:39]: I'd go for returning it, as it can be overlooked that there is a problem, by having the warnings not visible.

**Review by ntsekouras — APPROVED** [review]
Great work @mcsf ! In a follow up this hook can be moved and reused by other such blocks! 🚀

File: packages/block-library/src/block/index.php
Comment (gziolo) [reply packages/block-library/src/block/index.php:39]: I don't know if that is helpful for someone browsing the website. It's not that far to displaying an error when PHP can't connect with the database. You could display it maybe when `WP_DEBUG` is enabled but maybe with a ton of surrounding red colors 😄

**Comment by sirreal** [comment]
Thanks for working on this, the experience is very poor when recursion causes crashes right now.

File: packages/block-library/src/block/index.php
Comment (mcsf) [reply packages/block-library/src/block/index.php]: I went with `WP_DEBUG_DISPLAY`, per:

https://github.com/WordPress/wordpress-develop/blob/21b667b8405d8934c0ed8463e3273f3e4c22c989/src/wp-includes/load.php#L394-L400

**Comment by mcsf** [comment]
Thanks for all the great feedback! I will let the automated tests run, and allow about an hour for any last observations before I merge this.

**Comment by mcsf** [comment]
Opened #28428 as a follow-up.

**Comment by gziolo** [comment]
I started playing with applying the same technique for the Template Part block that is a bit more tricky to handle as it doesn't provide a unique identifier. Anyway, when going through the PHP implementation I was worried that the change proposed prevents rendering the same Reusable block multiple times in one editor. I was able to confirm that it doesn't work, see:

_Editor:_
Screen Shot 2021-01-25 at 10 34 39

_Front-end:_
Screen Shot 2021-01-25 at 10 34 56


While it works perfectly fine in the editor because it is using Context API, in PHP it uses `static` variable that is shared between all function calls so it will consider any subsequent render to the same block as invalid because it isn't aware of the concepts of rendering tree.

**Comment by carlomanf** [comment]
> Anyway, when going through the PHP implementation I was worried that the change proposed prevents rendering the same Reusable block multiple times in one editor. I was able to confirm that it doesn't work, see:

#26951 solved this by [unsetting the block from the static variable after the end of the render.](https://github.com/WordPress/gutenberg/pull/26951/files#diff-8372ba798df02239c3dd7da6a3d082bd1b57bebd7f6daf746412b00c709922cfR256) It should be an easy fix I think.

**Comment by ockham** [comment]
Thanks @gziolo for bringing this up!

I'm a bit surprised that the issue wasn't caught by this test :thinking: https://github.com/WordPress/gutenberg/blob/437a546a427ef39b9a3af61b33576bba84160940/packages/e2e-tests/specs/editor/various/reusable-blocks.test.js#L295

**Comment by ockham** [comment]
> Thanks @gziolo for bringing this up!
> 
> I'm a bit surprised that the issue wasn't caught by this test 
> 
> https://github.com/WordPress/gutenberg/blob/437a546a427ef39b9a3af61b33576bba84160940/packages/e2e-tests/specs/editor/various/reusable-blocks.test.js#L295

Ah, disregard. The e2e test checks only the editor, not the frontend (PHP).

**Comment by mcsf** [comment]
> Anyway, when going through the PHP implementation I was worried that the change proposed prevents rendering the same Reusable block multiple times in one editor. I was able to confirm that it doesn't work […]

Thanks for the report! Opened a fix at #28461.

**Pull request merged**
